### PR TITLE
FCFs: require proc types to have named formals, support throws keyword

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
FCFs: require proc types to have named formals, support throws keyword

The meaning of an unannotated formal (e.g., `x` in `proc(x)`) in a
type context is currently unclear. Turn such cases into errors
until the meaning is decided (see #20579 to participate).

Also make it an error to use an explicitly anonymous formal in a
procedure definition, as all formals must have names.

```chapel
// Error 1: meaning of an unannotated formal is ambiguous
type T = proc(int, int): int;

// Error 2: formals in procedure definitions must have names
proc foo(_: int, _: int) {}
```

While here, adjust the parser and implementation for FCFs to support
the `throws` keyword and throwing functions. Retroactively adjust
the productions for the `lambda` keyword to support throwing as well.
This was relatively trivial.

Also move the implementations of specialized errors in `dyno` into
their own source file `error-classes-list.cpp`, which matches the
name of the header where they are all declared.

Reviewed by @DanilaFe. Thanks!

TESTING

- [x] `linux64`, `standard`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
@DanilaFe Would you be willing to review my modifications to the DIAGNOSTIC_CLASS macro and my moving around the subclasses of `ErrorBase`?

File: compiler/dyno/include/chpl/framework/error-classes-list.h [hunk 1]
```diff
@@ -38,6 +38,18 @@ ERROR_CLASS(PrototypeInclude, const uast::Include*, const uast::Module*)
...
+    const uast::Variable*,
+    std::string,
+    std::string)
 WARNING_CLASS(ImplicitSubModule, const uast::Module*, UniqueString)
-WARNING_CLASS(ImplicitFileModule, const uast::AstNode*, const uast::Module*, const uast::Module*)
+WARNING_CLASS(ImplicitFileModule,
+    const uast::AstNode*,
+    const uast::Module*,
+    const uast::Module*)
+ERROR_CLASS(ProcTypeUnannotatedFormal,
```
Comment (DanilaFe) [inline compiler/dyno/include/chpl/framework/error-classes-list.h:50]: I've been deliberately putting the warnings below the errors, but I don't know if that's good ordering. What are your thoughts?

File: compiler/dyno/lib/framework/error-classes-list.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+void ErrorProcTypeUnannotatedFormal::write(ErrorWriterBase& wr) const {
+  auto sig = std::get<const uast::FunctionSignature*>(info);
+  auto formal = std::get<const uast::AnonFormal*>(info);
+  wr.heading(kind_, formal, "unannotated formal is ambiguous in this "
+                            "context");
+  wr.code(sig, {formal});
+  wr.message("The meaning of an unannotated formal (a formal not of the "
+             "form '<name>:<type>') in a procedure type is ambiguous."
+             "It is currently undecided whether it represents a formal's "
```
Comment (DanilaFe) [inline compiler/dyno/lib/framework/error-classes-list.cpp:180]: Isn't this missing a space after '.'?

File: compiler/dyno/lib/uast/post-parse-checks.cpp [hunk 3]
```diff
@@ -653,6 +656,27 @@ void Visitor::checkLambdaReturnIntent(const Function* node) {
   }
 }
 
+void
+Visitor::checkProcTypeFormalsAreAnnotated(const FunctionSignature* node) {
+  bool isProcType = (parent() && parent()->isFunction()) ? false : true;
```
Comment (DanilaFe) [inline compiler/dyno/lib/uast/post-parse-checks.cpp:661]:
```suggestion
  bool isProcType = !(parent() && parent()->isFunction());
```
Which also means you can do:
```suggestion
  bool isProcType = !parent() || !parent()->isFunction();
```

File: test/functions/fcf/ExplicitAnonFormalInNamedProc.chpl [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+
+
```
Comment (DanilaFe) [inline test/functions/fcf/ExplicitAnonFormalInNamedProc.chpl:3]: Do we need this white space?

File: test/functions/fcf/ExplicitAnonFormalInNamedProc.chpl [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+proc test() {
+  proc foo(_: int, _: int) {}
+}
+test();
+
```
Comment (DanilaFe) [inline test/functions/fcf/ExplicitAnonFormalInNamedProc.chpl:8]: And here?

File: test/functions/fcf/UnannotatedFormalInProcType.chpl [hunk 6]
```diff
@@ -0,0 +1,9 @@
+
+
+
```
Comment (DanilaFe) [inline test/functions/fcf/UnannotatedFormalInProcType.chpl:3]: And here?

**Review by DanilaFe — APPROVED** [review]

File: compiler/dyno/include/chpl/framework/error-classes-list.h (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/framework/error-classes-list.h:50]: I don't think an ordering mattered and would rather just do FCFS, but it's up to you I guess

File: compiler/dyno/lib/framework/error-classes-list.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/dyno/lib/framework/error-classes-list.cpp:180]: Sure is

File: compiler/dyno/lib/uast/post-parse-checks.cpp (hunk 3 above)
Comment (dlongnecke-cray) [reply compiler/dyno/lib/uast/post-parse-checks.cpp:661]: "Indeed"

File: compiler/dyno/include/chpl/framework/error-classes-list.h (hunk 1 above)
Comment (DanilaFe) [reply compiler/dyno/include/chpl/framework/error-classes-list.h:50]: Keep it as is then :)

File: compiler/dyno/lib/parsing/chpl.ypp [hunk 7]
```diff
@@ -2262,11 +2263,11 @@ lambda_decl_expr:
       fp.formals = $2;
       fp.returnIntent = $3;
       fp.returnType = $4;
-      WhereAndLifetime wl = $5;
-      fp.body = $7;
-
+      fp.throws = ($5 == ThrowsTag_THROWS) ? true : false;
```
Comment (DanilaFe) [inline compiler/dyno/lib/parsing/chpl.ypp:2266]: why not `$5 == ThrowsTag_THROWS`?

File: test/functions/lambda/lambdaThrows.chpl [hunk 8]
```diff
@@ -0,0 +1,8 @@
+const f1 = lambda(n: int): int throws { return n*n; };
+const f2 = lambda(n: int): int throws { throw new Error(n:string); };
+
+proc test() {
+  try! writeln(f1(4));
```
Comment (DanilaFe) [inline test/functions/lambda/lambdaThrows.chpl:5]: Why written in one but not the other?

**Review by DanilaFe — APPROVED** [review]

File: compiler/dyno/lib/parsing/chpl.ypp (hunk 7 above)
Comment (dlongnecke-cray) [reply compiler/dyno/lib/parsing/chpl.ypp:2266]: lolol it's true it's true, but in this case i just copy/pasted an existing line

**Pull request merged**